dali_tf_plugin/dali_dataset_op.cc

```diff
@@ -61,46 +61,50 @@ namespace {
 
 class DALIDatasetOp : public DatasetOpKernel {
  public:
-  explicit DALIDatasetOp(OpKernelConstruction* context) 
-    : DatasetOpKernel(context),
-      is_gpu_device_(context->device_type() == "GPU") { 
-      OP_REQUIRES_OK(context, context->GetAttr("pipeline", &pipeline_));
-      OP_REQUIRES_OK(context, context->GetAttr("batch_size", &batch_size_));
-      OP_REQUIRES_OK(context, context->GetAttr("num_threads", &num_threads_));
-      OP_REQUIRES_OK(context, context->GetAttr("device_id", &device_id_));
-      OP_REQUIRES_OK(context, context->GetAttr("exec_separated", &exec_separated_));
-      OP_REQUIRES_OK(context, context->GetAttr("prefetch_queue_depth", &prefetch_queue_depth_));
-      OP_REQUIRES_OK(context, context->GetAttr("cpu_prefetch_queue_depth", &cpu_prefetch_queue_depth_));
-      OP_REQUIRES_OK(context, context->GetAttr("gpu_prefetch_queue_depth", &gpu_prefetch_queue_depth_));
-      OP_REQUIRES_OK(context, context->GetAttr("shapes", &shapes_));
-      OP_REQUIRES_OK(context, context->GetAttr("dtypes", &dtypes_));
+  explicit DALIDatasetOp(OpKernelConstruction *context)
+      : DatasetOpKernel(context), is_gpu_device_(context->device_type() == "GPU") {
+    FillPipelineDef(context, pipeline_def_);
+    OP_REQUIRES_OK(context, context->GetAttr("shapes", &shapes_));
+    OP_REQUIRES_OK(context, context->GetAttr("dtypes", &dtypes_));
+  }
+
+    void MakeDataset(OpKernelContext *context, DatasetBase **output) override {
+      *output = new Dataset(context, pipeline_def_, shapes_, dtypes_, is_gpu_device_);
     }
 
-  void MakeDataset(OpKernelContext* context, DatasetBase** output) override {
-    *output = new Dataset(
-      context, 
-      pipeline_, 
-      batch_size_, 
-      num_threads_,
-      device_id_,
-      exec_separated_,
-      prefetch_queue_depth_,
-      cpu_prefetch_queue_depth_,
-      gpu_prefetch_queue_depth_,
-      shapes_, 
-      dtypes_,
-      is_gpu_device_);
+ private:
+  struct PipelineDef {
```

Comment:
Just a thought: maybe this struct is more general and we could put it in DALI rather than in Dataset? Maybe somwhere in C API or something?

Reply:
To put it in C API we would need to get rid of string, but it seems doable.
Maybe I will pursue it in some follow up.